### PR TITLE
Add language model support

### PR DESCRIPTION
**Pull request by ejnnr**
Goal for this PR is to be able to train language models on simple classification tasks, and be able to extract their activations. Actual anomaly detection tasks (like backdoors or measurement tampering) will probably be separate PRs, same for any adaptations that existing detectors will need.

Running
```bash
python -m cupbearer.scripts.train_classifier --train_data imdb --model transformer --batch_size 24
```
already runs, but there's lots of stuff to do:
- [x] There might be a memory leak right now, usage goes up significantly over the course of training. Maybe related to https://github.com/neelnanda-io/TransformerLens/issues/290, maybe we just need to call `torch.cuda.empty_cache()`, or maybe there's a memory leak somewhere in `cupbearer` that I just didn't notice before because all the models where tiny (or maybe the new model specifically has a leak). Can also try training models directly with huggingface to narrow down the cause.
- [ ] I'm not sure training actually works, after 1 epoch accuracy is still 50%, and I haven't trained longer because of the memory leak.
- [ ] Check `eval_classifier`, it'll probably need some small adjustments
- [ ] Check that using the new model in tasks actually works (can just use a dummy task for tests where untrusted data comes from the clean distribution as well)
- [ ] Automated smoke tests
- [ ] Make things configurable (e.g. the output_dim of 2 is basically just hard-coded right now)
- [ ] I'm not sure I like the new `DataFormat`. If we do keep it, we should consider merging in `num_classes` etc. as well (i.e. have a single object for configuring both input and output format)
- [ ] Test that it works for models other than `pythia-14m`

### CODE REVIEW

**Comment by VRehnberg** [comment]
I'm not seeing a memory leak and accuracy seems to go up (albeit, very slowly).
 - https://api.wandb.ai/links/vrehnberg/nb1wml6l
 - https://scruffy.c3se.chalmers.se/d/alvis-job/alvis-job?var-jobid=1586128&from=1700807940000&orgId=1&to=1700808660000

**Comment by ejnnr** [comment]
Thanks, that's good to know re memory leak! I was running on MPS and it seems like that might be the problem. If I run on (M1) CPU, memory usage stays constant, but apparently at ~450 MB, which I think is way too low, so I'm not sure how much I trust that. I'll look into fixing this a bit more anyway, since it would be nice to sometimes run small tests locally, but if it works on CUDA, that's the most important thing.

My sense would have been that we should see better accuracy/loss after training for 10 epochs, so I still suspect something's wrong there.

**Comment by ejnnr** [comment]
Some progress notes on the memory leak: I've narrowed this down to only happening when the sequence lengths change between batches. It also doesn't have anything to do with TransformerLens, cupbearer, or pytorch lightning, I could reproduce it with only huggingface and pytorch.

Below is a script for reproducing this. With `LEAK = False`, driver MPS memory maxes out at around 2300 MB for me, and memory according to activity monitor also stays constant. With `LEAK = True`, memory usage according to activity monitor tends to increase (to at least 4600 MB, stopped it there), more variable according to `torch`. This is despite the `torch.mps.empty_cache()` call.

Seems like a bug in either pytorch or huggingface. We can pretty easily work around it by using fixed size batches if we're on MPS, but might also be good to dig into this and actually fix it.

```python
LEAK = False

import torch
import random
from transformers import AutoTokenizer, AutoModel
import datasets

model_name = "EleutherAI/pythia-14m"
tokenizer = AutoTokenizer.from_pretrained(model_name)
tokenizer.pad_token = tokenizer.eos_token
# We don't want logits, so use AutoModel instead of AutoModelForCausalLM
model = AutoModel.from_pretrained(model_name)
device = torch.device("mps")
model.to(device)

optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

dataset = datasets.load_dataset("imdb", split="train")

BATCH_SIZE = 16

for i in range(0, 1000):
    optimizer.zero_grad()
    indices = random.sample(range(len(dataset)), BATCH_SIZE)

    inputs = dataset[indices]["text"]
    labels = dataset[indices]["label"]

    if LEAK:
        tokenizer_res = tokenizer(
            inputs,
            truncation=True,
            padding="longest",
            return_tensors="pt",
            max_length=512,
        )
    else:
        tokenizer_res = tokenizer(
            inputs,
            truncation=True,
            padding="max_length",
            return_tensors="pt",
            max_length=512,
        )
    output = model(
        input_ids=tokenizer_res["input_ids"].to(device),
        attention_mask=tokenizer_res["attention_mask"].to(device),
        use_cache=False,
    )

    # Just arbitrarily pick out some hidden dimensions.
    # Taking the first position is obviously silly but this is for debugging.
    logits = output.last_hidden_state[:, 0, :2]

    loss = torch.nn.functional.cross_entropy(
        logits, torch.tensor(labels, dtype=torch.long, device=device)
    )

    loss.backward()
    optimizer.step()
    if i % 10 == 0:
        print(
            f"Step {i}: {loss}, {torch.mps.driver_allocated_memory() / 1024 ** 2}, {torch.mps.current_allocated_memory() / 1024 ** 2}"
        )
```
